modules/core/src/lib/layers/selection-layer.js

```diff
@@ -0,0 +1,272 @@
+
+    if (landPoints.length !== 2) return;
+
+    const [x1, y1] = mousePoints[0];
+    const [x2, y2] = mousePoints[1];
+    const pickingInfos = this.context.layerManager.pickObjects({
+      viewports: [this.context.viewport],
+      x: Math.min(x1, x2),
+      y: Math.min(y1, y2),
+      width: Math.abs(x2 - x1),
+      height: Math.abs(y2 - y1),
+      layerIds
+    });
+
+    onSelection({ pickingInfos });
+  }
+
+  _selectPolygonObjects() {
+    const { layerIds, onSelection } = this.props;
+    const { mousePoints } = this.state;
+
+    const allX = mousePoints.map(mousePoint => mousePoint[0]);
+    const allY = mousePoints.map(mousePoint => mousePoint[1]);
+    const x = Math.min(...allX);
+    const y = Math.min(...allY);
+    const maxX = Math.max(...allX);
+    const maxY = Math.max(...allY);
+
+    const pickingInfos = this.context.layerManager.pickObjects({
+      viewports: [this.context.viewport],
+      x,
+      y,
+      width: maxX - x,
+      height: maxY - y,
+      layerIds: [LAYER_ID_PICK, ...layerIds]
+    });
+
+    onSelection({
+      // TODO: fix this
+      pickingInfos: pickingInfos.filter(item => item.layer.id !== LAYER_ID_PICK)
+    });
+  }
+
+  onClick({ groundCoords, screenCoords }: ClickEvent) {
+    const { mousePoints, landPoints, validPolygon } = this.state;
+
+    if (this.props.selectionType === SELECTION_TYPE.POLYGON) {
+      // If landPoints.length is zero we need to insert two points.
+      if (landPoints.length === 0) {
+        this.setState({
+          landPoints: [groundCoords, groundCoords],
+          mousePoints: [screenCoords, screenCoords]
+        });
+      } else if (this.state.landPoints.length < 3 || validPolygon) {
+        // Don't insert if polygon is invalid
+        this.setState({
+          landPoints: [...landPoints, groundCoords],
+          mousePoints: [...mousePoints, screenCoords]
+        });
+      }
+    }
+  }
+
+  onDoubleClick(event: DoubleClickEvent) {
+    const { landPoints, validPolygon } = this.state;
+    if (
+      this.props.selectionType === SELECTION_TYPE.POLYGON &&
+      landPoints.length > 4 &&
+      validPolygon
+    ) {
+      this._selectPolygonObjects();
+      this.reset();
+    }
+  }
+
+  onStartDragging({ groundCoords, screenCoords }: StartDraggingEvent) {
+    if (this.props.selectionType === SELECTION_TYPE.RECTANGLE) {
+      this.setState({
+        landPoints: [groundCoords, groundCoords],
+        mousePoints: [screenCoords, screenCoords]
+      });
+    }
+  }
+
+  reset() {
+    this.setState({
+      landPoints: [],
+      mousePoints: []
+    });
+  }
+
+  onStopDragging(event: StopDraggingEvent) {
+    if (this.props.selectionType === SELECTION_TYPE.RECTANGLE) {
+      this._selectRectangleObjects();
+      this.reset();
+    }
+  }
+
+  onPointerMove({ groundCoords, screenCoords, sourceEvent }: PointerMoveEvent) {
+    if (this.props.selectionType !== SELECTION_TYPE.NONE && this.state.landPoints.length) {
```

Comment:
Looks like the map pans a bit before it calls `stopPropagation`. Look at `onPointerMove` handling in `ModifyHandler` to see how I call `stopPropagation` between when pointer went down and when drag is detected.